rfcs/000-caller-side-managed-buffers.md

```diff
@@ -0,0 +1,833 @@
+        // ..
+
+        conn.discard_early_data();
+    }
+}
+```
+
+An alternative to this is to, for example, include the early data in the `State` variant itself.
+
+```rust
+enum State<'a> {
+    ReceivedEarlyData(&'a [u8]),
+}
+```
+
+But this alternative has two issues:
+`State`, and thus `Status` as well, would have to gain a lifetime that ties it to the `incoming_tls` buffer.
+Given how Rust works, even if `State` was not in the `ReceivedEarlyData` variant, the `status.state` field would still "freeze" `incoming_tls` meaning that it can't be mutated until `status.state` goes out of scope / is dropped.
+
+The other issues is that if `process_tls_records` automatically decrypts early data, the end-user no longer has the option to discard the early data without decrypting it first (see snippet below) to save some CPU work.
+
+```rust
+// this is possible with the API in the main proposal
+
+if status.received_early_data() {
+    conn.discard_early_data();
+    // without decrypting it first
+}
+```
+
+## "Less" allocations in `State`
+
+On the other hand, some of the `State` variants include owned data types like `Vec<u8>`.
+These are copies of, potentially encrypted, data that was stored in the `incoming_tls` buffer.
+
+It would be possible to avoid some allocations in `LlConnection` by turning the enum variants into unit structs and adding new methods to `LlConnection` to yield slices to the relevant part of `incoming_tls`. For example,
+
+```rust
+struct CertificateRef<'a>(pub &'a [u8]);
+
+struct CertificateEntryRef<'a> {
+    pub cert: CertificateRef<'a>,
+    pub exts: Vec<CertificateExtension>,
+}
+
+loop {
+    // ..
+
+    match status.state {
+        State::ReceivedCertificate => {
+            let certificates: Vec<CertificateEntryRef> = conn.decrypt_certificate(&mut incoming_tls);
+            // ..
+        }
+
+    }
+}
+```
+
+Given that `certificates` freezes `incoming_tls`, the end-user will likely have to copy the data into an owned version (`Certificate`) so
+they can stash the certificates for use in a later iteration of the event loop or to send them to a separate async task.
+In the end, the allocation is not eliminated but simply made more explicit, as the end user will have to perform it.
+
+This main proposal's criteria for whether to include owned data into one of `State` variants is to re-use existing owned types and avoid creating new "ref" versions of them.
+For instance, the internal `Codec` trait which creates instances the `Certificate` type from a `Reader` slice can only create owned types, not slices.
+
+# Unresolved questions
+
+what should the behavior be in these scenarios
+
+- `Status.state` is `State::MustTransmitTlsData`, `outgoing_tls` is not fully transmitted and `process_tls_records` is called
+  - should `wants_write` "latch" to the `true` value until `outgoing_tls` is observed as being empty?
+- `Capabilities.may_send_app_data` returns `false` and `encrypt_outgoing` is called
```

Comment:
As mentioned previously, can we statically make this impossible?

Reply:
yes, `sketch2` achieves this. `may_send_app_data` has been turned into a `State` variant that yields a newtype over `LlConnectionCommon` and provides the `encrypt_outgoing` method